_posts/2018-01-19-deep-dive-into-database-timeouts-in-rails.md

```diff
@@ -0,0 +1,355 @@
+~~~
+
+After `read_timeout` has elapsed, we see that a new connection is established to the database, and the first one has transitioned to a `FIN-WAIT` state. This new TCP connection is in the `ESTAB` state (line 3), because we have only stopped the database on the application layer, but the sockets to the container still accept the TCP handshake on the transport layer.
+
+~~~ bash
+$ ss -tnp
+FIN-WAIT-2 0      0               172.18.0.4:54304          172.18.0.3:3306
+ESTAB      0      0               172.18.0.4:54308          172.18.0.3:3306   users:(("ruby",pid=11683,fd=13))
+ESTAB      0      0               172.18.0.4:3000           172.18.0.1:60878  users:(("ruby",pid=11683,fd=12))
+~~~
+
+After `connect_timeout` has elapsed, the request terminates with a 500 error, and we observe that all the connections are in the `FIN-WAIT` state.
+
+~~~ bash
+$ ss -tnp
+State      Recv-Q Send-Q       Local Address:Port         Peer Address:Port
+FIN-WAIT-2 0      0               172.18.0.4:54310          172.18.0.3:3306
+FIN-WAIT-2 0      0               172.18.0.4:54304          172.18.0.3:3306
+FIN-WAIT-2 0      0               172.18.0.4:54308          172.18.0.3:3306
+~~~
+
+The experimental data can be found [below](#experimental-data).
+
+#### Findings
+
+It’s worth noting that when setting `connect_timeout` and `read_timeout` in the `database.yml`, there is a difference between empty values and the case where the key is missing entirely in the file. If the values are empty, scenario 1 will fail to terminate after 5 minutes, but if the keys are absent, scenario 1 will fail after 120 seconds, which is the default for `connect_timeout`.
+
+##### Experiment 1 Findings
+
+The request waits for `connect_timeout` to connect to the database, where the default value (when not specified) is indeed 120 seconds.
+
+As expected, connecting to the database with no existing connections is independent of the `read_timeout`.
+
+##### Experiment 2 Findings
+
+The request waits for `read_timeout` + `connect_timeout` before failing. This is because the connection pool waits for `read_timeout` on the existing connection before terminating it, and then waits for `connect_timeout` as it tries to establish a new connection to `db_other`.
+
+## Analysis
+
+With these findings, we can try to understand how the lack of these timeouts affected our Rails server in production during and after the database failover.
+
+### Establishing Terms
+
+Our application server constantly receives requests, out of which a certain percentage of requests will trigger the code to connect to the affected database, which we'll call _x_-type requests. The other requests, that do not trigger a database connection, we'll call _x’_-type requests.
+
+### Analysis
+
+With the background knowledge gathered in our experiments, let’s try to analyse all the steps that happened during our production outage.
+
+1. Rails started from a clean state, with no connections set up to the database initially
```

Comment:
Use 1. throughout for numbered bullet points in Markdown so that they increment automatically

Reply:
👍 Ah I did not know this, thanks for the tip!